value/valueunstructured.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package value
+
+import (
+	"fmt"
+	"sync"
+)
+
+var viPool = sync.Pool{
+	New: func() interface{} {
+		return &valueUnstructured{}
+	},
+}
+
+// NewValueInterface creates a Value backed by an "interface{}" type,
+// typically an unstructured object in Kubernetes world.
+func NewValueInterface(v interface{}) Value {
+	vi := viPool.Get().(*valueUnstructured)
+	vi.Value = v
+	return Value(vi)
+}
+
+type valueUnstructured struct {
+	Value interface{}
+}
+
+func (v valueUnstructured) IsMap() bool {
+	if _, ok := v.Value.(map[string]interface{}); ok {
+		return true
+	}
+	if _, ok := v.Value.(map[interface{}]interface{}); ok {
+		return true
+	}
+	return false
+}
+
+func (v valueUnstructured) Map() Map {
+	if v.Value == nil {
+		return mapUnstructuredString(nil)
```

Comment:
Why this and not just `nil`? e.g. mapUnstructuredString's Get doesn't look like it's prepared to deal with this.

I think `panic`ing is an option, too--that'd be like the built-in reflect package.

Reply:
You're right, that doesn't make a mistake. There are a few places where we treat nil as a map so I don't think we can panic here.

Reply:
We should fix that for real-- I'd honestly prefer to make people always check IsNull first.

Panic here and exhaustively test?

None of the other types here permit nils.

Reply:
Yeah, it crashes in union, should be fairly limited and easy to test, will fix now

Reply:
solved